File: pagmoSharp/pagmoExtensions/Algorithms/AlgorithmInterop.cs
using System;

namespace pagmo
{
    internal static class AlgorithmInterop
    {
        internal static algorithm NormalizeToTypeErased(IAlgorithm source)
        {
            if (source is null)
            {
                throw new ArgumentNullException(nameof(source));
            }

            return source switch
            {
                algorithm typeErased => typeErased,
                bee_colony beeColony => beeColony.to_algorithm(),
                cmaes covarianceMatrixAdaptationEvolutionStrategy => covarianceMatrixAdaptationEvolutionStrategy.to_algorithm(),
                compass_search compassSearch => compassSearch.to_algorithm(),
                ihs improvedHarmonySearch => improvedHarmonySearch.to_algorithm(),
                nsga2 nsga2Algorithm => nsga2Algorithm.to_algorithm(),
                nlopt nloptAlgorithm => nloptAlgorithm.to_algorithm(),
                moead moeadAlgorithm => moeadAlgorithm.to_algorithm(),
                moead_gen moeadGenerationalAlgorithm => moeadGenerationalAlgorithm.to_algorithm(),
                maco multiObjectiveAntColonyOptimization => multiObjectiveAntColonyOptimization.to_algorithm(),
                mbh monotonicBasinHopping => monotonicBasinHopping.to_algorithm(),
                cstrs_self_adaptive selfAdaptiveConstraints => selfAdaptiveConstraints.to_algorithm(),
                de differentialEvolution => differentialEvolution.to_algorithm(),
                de1220 differentialEvolution1220 => differentialEvolution1220.to_algorithm(),
                gaco generalizedAntColonyOptimization => generalizedAntColonyOptimization.to_algorithm(),
                gwo greyWolfOptimizer => greyWolfOptimizer.to_algorithm(),
                nspso nonDominatedSortingParticleSwarmOptimizer => nonDominatedSortingParticleSwarmOptimizer.to_algorithm(),
                null_algorithm nullAlgorithm => nullAlgorithm.to_algorithm(),
                pso particleSwarmOptimizer => particleSwarmOptimizer.to_algorithm(),
                pso_gen particleSwarmOptimizerGenerational => particleSwarmOptimizerGenerational.to_algorithm(),
                sade selfAdaptiveDifferentialEvolution => selfAdaptiveDifferentialEvolution.to_algorithm(),
                sea simpleEvolutionaryAlgorithm => simpleEvolutionaryAlgorithm.to_algorithm(),
                sga simpleGeneticAlgorithm => simpleGeneticAlgorithm.to_algorithm(),
                simulated_annealing simulatedAnnealing => simulatedAnnealing.to_algorithm(),
                xnes exponentialNaturalEvolutionStrategy => exponentialNaturalEvolutionStrategy.to_algorithm(),
                _ => new algorithm(source)
            };
        }
    }
}

File: java/core/src/main/java/io/github/samthegliderpilot/pagmo4j/algorithms/AlgorithmInterop.java
package io.github.samthegliderpilot.pagmo4j.algorithms;

import io.github.samthegliderpilot.pagmo4j.*;
import io.github.samthegliderpilot.pagmo4j.NativeInterop;

/**
 * Normalizes any {@link IAlgorithm} to a type-erased {@link algorithm} for
 * pagmo's island/archipelago machinery.
 *
 * <p>Concrete pagmo algorithm classes are passed through their native
 * {@code to_algorithm()} conversion path. Custom managed algorithms are
 * wrapped via the callback bridge.
 */
public final class AlgorithmInterop {

    private AlgorithmInterop() {}

    /**
     * Normalizes any {@link IAlgorithm} to a type-erased {@link algorithm}.
     *
     * <p>Concrete pagmo algorithm classes (e.g. {@code de}, {@code sade}) are converted through
     * their native {@code to_algorithm()} path for zero-copy wrapping. Custom managed
     * {@link IAlgorithm} implementations are wrapped via the director callback bridge.
     *
     * <p><b>NOTE:</b> This method must be updated whenever a new concrete algorithm class is
     * added to the pagmo4j binding. Each concrete type needs an explicit {@code instanceof}
     * branch so it uses the native conversion path rather than the slower director path.
     */
    public static algorithm normalizeToTypeErased(IAlgorithm source) {
        if (source == null) throw new NullPointerException("source");

        if (source instanceof algorithm)            return (algorithm) source;
        if (source instanceof bee_colony)           return ((bee_colony) source).to_algorithm();
        if (source instanceof cmaes)                return ((cmaes) source).to_algorithm();
        if (source instanceof compass_search)       return ((compass_search) source).to_algorithm();
        if (source instanceof ihs)                  return ((ihs) source).to_algorithm();
        if (source instanceof ipopt)                return ((ipopt) source).to_algorithm();
        if (source instanceof nsga2)                return ((nsga2) source).to_algorithm();
        if (source instanceof nlopt)                return ((nlopt) source).to_algorithm();
        if (source instanceof moead)                return ((moead) source).to_algorithm();
        if (source instanceof moead_gen)            return ((moead_gen) source).to_algorithm();
        if (source instanceof maco)                 return ((maco) source).to_algorithm();
        if (source instanceof mbh)                  return ((mbh) source).to_algorithm();
        if (source instanceof cstrs_self_adaptive)  return ((cstrs_self_adaptive) source).to_algorithm();
        if (source instanceof de)                   return ((de) source).to_algorithm();
        if (source instanceof de1220)               return ((de1220) source).to_algorithm();
        if (source instanceof gaco)                 return ((gaco) source).to_algorithm();
        if (source instanceof gwo)                  return ((gwo) source).to_algorithm();
        if (source instanceof nspso)                return ((nspso) source).to_algorithm();
        if (source instanceof null_algorithm)       return ((null_algorithm) source).to_algorithm();
        if (source instanceof pso)                  return ((pso) source).to_algorithm();
        if (source instanceof pso_gen)              return ((pso_gen) source).to_algorithm();
        if (source instanceof sade)                 return ((sade) source).to_algorithm();
        if (source instanceof sea)                  return ((sea) source).to_algorithm();
        if (source instanceof sga)                  return ((sga) source).to_algorithm();
        if (source instanceof simulated_annealing)  return ((simulated_annealing) source).to_algorithm();
        if (source instanceof xnes)                 return ((xnes) source).to_algorithm();

        return new algorithm(source);
    }
}
